README.md
# rana
yet another anagram algorithm, this time in Rust

## Usage

The text provided by `--help-long`.
```
USAGE:
    rana [FLAGS] [OPTIONS] <word>

FLAGS:
    -w, --words-in     Returns the set of words composable from the letters in the input phrase
        --strict       When finding --words-in, returns only words that occur in some anagram
        --prove        Like --strict, but emits a phrase proving this word occurs in an anagram.
    -h, --help         Prints help information
        --help-long    Prints *detailed* help information
    -C, --no-cache     Do not cache partial results (this saves memory and costs speed)
    -r, --random       (Partially) shuffle order of discovery
        --ribbit       Ego sum
    -V, --version      Prints version information

OPTIONS:
    -d, --dictionary <file>          A line-delimited list of words usable in anagrams [default: ~/.anagram-dictionary.txt]
    -x, --exclude <word>...          Exclude this word from anagrams
    -i, --include <word>...          Include this word in the anagrams
    -l, --limit <n>                  Only find this many anagrams
    -m, --minimum-word-length <n>    Words in anagrams must be at least this long
    -t, --threads <n>                The number of threads to use during anagram collection [default: 8]

ARGS:
    <word>...    The words for which you want an anagram

Rana generates all the possible anagrams from a given phrase and
dictionary. Note "given some dictionary." Rana does not have a word list
built in. You must tell it what words it may use in an anagram. I have made
myself such a list out of a list of English words I found on the Internet from
which I delted all the words likely to offend people. By default rana will
look in your home directory for a file called .anagrams-dictionary.txt.

In many cases a simple phrase will have hundreds of thousands or millions of
anagrams, setting aside permutations. The phrase "rotten apple", for example,
with a fairly ordinary dictionary of of 109,217 English words, produces 2695
anagrams. Here are 10:

    pone prattle
    plea portent
    pole pattern
    portent pale
    platter pone
    planter poet
    potent paler
    porn palette
    pron palette
    poler patent

Because so many anagrams are available, you are likely to want to focus your
search. Rana provides several options to facilitate this.

--words-in

This will list all the words in your dictionary composable from some subset of
your phrase. You likely want to add --strict to this, so you only get words that
occur in *some* anagram. The --strict version is slower. If you want to verify
that each word listed has some anagram, you can add --prove, or replace --strict
with --prove. This will cause rana to emit the remaining words in an anagram
using the word in question.

--exclude

Discard from your word list particular words.

--include

Include only those phrases which include particular words.

--limit

Only provide a sample of this many phrases.

--random

Shuffle the search order over partial results while searching for anagrams. This
does not provide a fully random sample of the possible anagrams, since only the
results found at any point are shuffled, not all possible results, but this is
a decent way to look at a sample of anagrams when the phrase you've fed in has
many thousands of results. This is particularly useful when paired with --limit.

Caching and Threads

Rana by default uses as many processing threads as there are cores on your
machine. Generally this is what you want, but if you've got a lot of other
things going on, you can limit the number of available threads to reduce the
load your kernel has to deal with.

Rana also uses a dynamic programming algorithm to reduce the complexity of
finding algorithms for large phrases. This is probably unnecessary for short
phrases, though rana provides no lower limit. For larger phrases, like the
complete alphabet, the cache used by the dynamic programming algorithm may grow
so large that the process crashes. If you turn off the cache rana will use
a constant amount of memory, though it may take considerably longer to find all
anagrams.

Text Normalization

Rana attempts to strip away certain characters from your word list and all
other textual input, so it will treat "c-a-t" and " C A T " the same as "cat".
Here is the actual code that does this:

    pub fn normalize(word: &str) -> String {
        word.trim()
            .to_lowercase()
            .chars()
            .filter(|c| c.is_alphabetic())
            .collect::<String>()
    }

I have not tested what this will do for something like ß or Í. You may want to
normalize the text yourself before you give it to rana.

NOTE:

The caching algorithm treats character counts as long base-10 numbers. So, for
example, "cat" might be 111 if "c" is 100s, "a" is 10s, and "t" is 1s. Then
"cad" might be 1110 -- "d" is 1000s, and there is one of them, but there is no
"t", so the 1s place is 0. If you have a phrase, like

    "Dhrtaraashtra uvaaca, dharmakshetre kurukshetre samavetaa yuyutsavah"

which has 15 a's, the "a" count has to be represented as 5 -- 15 mod 10 is 5. In
other words, there isn't space in the a's column of the number for all the a's
in the phrase. It is unlikely that this will ever cause trouble, because for a
particular phrase you are unlikely to be encounter two sets of character counts
during processing that have different counts but the same code. Also, a phrase
this size will consume so much cache space that the process will probably crash
before you encounter this collision.

Another consideration with caching is that this scheme can only accommodate
alphabets up to 38 characters in size.
```

An example use:

```
~ $ rana eat
eat
a et
eta
ate
tea
```

## Installation

Rana is available on `crates.io`, so you can install it with a

    cargo install ranagrams

You can also clone or copy this project to your machine and run
`cargo build --release` in its directory. This will produce an executable called
`target/release/rana`. To use the executable you will need a word list. I have
not checked mine in. You may have a file named something like
`/usr/share/dict/words` on your computer. If so, this will work, though it may
have every individual letter in it along with other words, which means if you
use it as your dictionary that `c a t` will be an anagram of `cat`. You can find
word lists on-line if you simply search for "word list". The list will need to
have all the forms you are interested in. Rana cannot infer "cats" from "cat",
much less "brought" from "bring".

## History and Credits

I have made many variants of this anagram algorithm. This is the first in Rust.
It is more efficient and faster than any of the previous versions. I cannot say
that this is particularly good or idiomatic Rust. It is the fist significant
bit of Rust I've ever written. To the extent that this is good Rust, the credit
is entirely due to @TurkeyMcMac, who knows Rust much better than I do and could
generally tell me when I was doing something particularly stupid.


CHANGES.md
# Change Log

## 1.0 *2020-11-22*
* fixing deprecation warning in CLI code
* upgraded some dependencies
* added sample word list to repository
* declaring this stable
## 0.2.4
* changing deprecated `std::env::home_dir` to `dirs::home_dir`
* improved the efficiency of --strict and --prove
* added optimization step so anagram discovery uses a trie containing only words embeddable in the given phrase
## 0.2.3
* changed dependency versions


tests/cli.rs
extern crate ranagrams;

#[test]
fn foo() {
    assert!(true)
}


src/main.rs
extern crate ranagrams;
use factory::WorkerFun;
use ranagrams::cli;
use ranagrams::factory;
use ranagrams::trie::{Trie, TrieNodeBuilder};
use ranagrams::util::{normalize, ToDo, Translator};
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::Read;
use std::sync::atomic::Ordering;
use std::sync::Arc;
extern crate clap;
use clap::ArgMatches;
extern crate num_cpus;
use std::ops::Deref;
use std::process;
extern crate dirs;
use dirs::home_dir;

fn main() {
    // parse the options
    let cpus = num_cpus::get().to_string();
    let default_dir = if let Some(mut buf) = home_dir() {
        buf.push(".anagram-dictionary.txt");
        Some(buf.to_str().unwrap().to_string())
    } else {
        None
    };
    let options = cli::parse(&cpus, default_dir.as_ref().map(String::deref)).get_matches();
    if options.is_present("long-help") {
        cli::parse(&cpus, default_dir.as_ref().map(String::deref))
            .print_help()
            .ok();
        println!("\n\n{}", cli::long_help());
        process::exit(0)
    }
    if options.is_present("ribbit") {
        println!("\n{}", include_str!("../rana.txt"));
        process::exit(0)
    }
    let threads = if options.is_present("set") && !options.is_present("strict") {
        // only one thread will ever be used
        1
    } else {
        match usize::from_str_radix(options.value_of("threads").unwrap(), 10) {
            Err(why) => {
                eprintln!("error parsing thread count: {}\n\n{}", why, options.usage());
                process::exit(1)
            }
            Ok(threads) => threads,
        }
    };
    let use_limit = options.is_present("limit");
    let limit = if use_limit {
        match usize::from_str_radix(options.value_of("limit").unwrap(), 10) {
            Err(why) => {
                eprintln!(
                    "could not parse anagram limit: {}\n\n{}",
                    why,
                    options.usage()
                );
                process::exit(1)
            }
            Ok(limit) => limit,
        }
    } else {
        0
    };
    let min_word_length = if options.is_present("min") {
        match usize::from_str_radix(options.value_of("min").unwrap(), 10) {
            Err(why) => {
                eprintln!(
                    "could not parse minimum word length: {}\n\n{}",
                    why,
                    options.usage()
                );
                process::exit(1)
            }
            Ok(min) => {
                if min == 0 {
                    eprintln!(
                        "minimum word length must be positive\n\n{}",
                        options.usage()
                    );
                    process::exit(1)
                } else {
                    min
                }
            }
        }
    } else {
        1
    };
    let trie_word_length = if options.is_present("strict") || options.is_present("prove") {
        1
    } else {
        min_word_length
    };
    let mut trie = make_trie(&options, trie_word_length);

    // create initial character count
    let mut cc = trie
        .translator
        .count(&normalize(""))
        .expect("no luck with the char count");
    // add all the words to anagramize
    for word in options.values_of("phrase").unwrap() {
        if let Some(usizes) = trie.translator.translate(word) {
            if !cc.add(usizes) {
                dictionary_error(word, &trie)
            }
        } else {
            dictionary_error(word, &trie)
        }
    }
    // subtract the words to include
    let prefixed = options.is_present("include");
    let mut prefix = String::new();
    if prefixed {
        for word in options.values_of("include").unwrap() {
            if let Some(usizes) = trie.translator.translate(word) {
                match cc.subtract(usizes) {
                    Some((i, copy)) => {
                        let normalized = trie.translator.etalsnart(&copy).unwrap();
                        eprintln!(
                            "attempt to use unavailable character in {}:\n\n\t{}-->{}",
                            &normalized,
                            &normalized[0..i],
                            &normalized[i..]
                        );
                        process::exit(1)
                    }
                    None => {
                        prefix.push_str(word);
                        prefix.push(' ');
                    }
                }
            } else {
                dictionary_error(word, &trie)
            }
        }
    }
    cc.set_limits();
    trie = trie.optimize(cc.clone());

    if options.is_present("set") {
        if options.is_present("strict") || options.is_present("prove") {
            let prove = options.is_present("prove");
            let sort_key = Vec::with_capacity(0);
            // find candidate words -- these are embedded in the initial character count and thus
            // are the only ones that might be in an anagram
            let mut found: Vec<String> = trie
                .words_for(Arc::new(cc.clone()), &sort_key, &true)
                .into_iter()
                .map(|(chars, _)| trie.translator.etalsnart(&chars).unwrap())
                .collect();
            found.sort();
            let mut cache: HashMap<String, String> = HashMap::new();
            for word in found {
                if word.len() >= min_word_length {
                    if cache.contains_key(&word) {
                        println!("{}", word);
                        if prove {
                            println!("\t{}", cache.get(&word).unwrap());
                        }
                    } else {
                        if let Some(usizes) = trie.translator.translate(word.as_str()) {
                            // can we make a least one anagram with the remainder after we subtract this word?
                            let mut cc = cc.clone();
                            cc.subtract(usizes.to_vec());
                            let materials = vec![ToDo::seed(cc)];
                            let noah = Arc::new(AnagramFun { root: trie.clone() });
                            let (messages, kill_switch) =
                                factory::manufacture(threads, 3, materials, noah.clone());
                            if let Some(Some(done)) = messages.iter().next() {
                                kill_switch.store(true, Ordering::Relaxed);
                                println!("{}", word);
                                let s = if prove {
                                    word.clone() + " " + &trie.stringify(done)
                                } else {
                                    trie.stringify(done)
                                };
                                if prove {
                                    println!("\t{}", s);
                                } else {
                                    // only do this if we're not proving, because we aren't always saving the proof in the cache
                                    for ow in words_in_word(&word, noah.clone(), threads) {
                                        if ow.len() >= min_word_length
                                            && ow > word
                                            && !cache.contains_key(&ow)
                                        {
                                            cache.insert(ow, s.clone());
                                        }
                                    }
                                }
                                for other_word in s.as_str().split(" ") {
                                    if prove {
                                        if other_word.len() >= min_word_length && other_word > &word
                                        {
                                            cache.insert(String::from(other_word), s.clone());
                                        }
                                    } else {
                                        for ow in words_in_word(other_word, noah.clone(), threads) {
                                            if ow.len() >= min_word_length
                                                && ow > word
                                                && !cache.contains_key(&ow)
                                            {
                                                cache.insert(ow, s.clone());
                                            }
                                        }
                                    }
                                }
                            } else {
                                trie.remove(&usizes);
                            }
                        }
                    }
                }
            }
        } else {
            let sort_key = Vec::with_capacity(0);
            let mut found: Vec<String> = trie
                .words_for(Arc::new(cc), &sort_key, &true)
                .into_iter()
                .map(|(chars, _)| trie.translator.etalsnart(&chars).unwrap())
                .collect();
            found.sort();
            for word in found {
                println!("{}", word);
            }
        }
    } else {
        let mut count = 0;
        let materials = vec![ToDo::seed(cc)];
        let noah = Arc::new(AnagramFun { root: trie });
        let mine = noah.clone();
        let (messages, kill_switch) = factory::manufacture(threads, 3, materials, noah);
        for m in messages {
            if let Some(todo) = m {
                if prefixed {
                    print!("{}", prefix);
                }
                println!("{}", mine.root.stringify(todo));
                if use_limit {
                    count += 1;
                    if count == limit {
                        kill_switch.store(true, Ordering::Relaxed);
                        break;
                    }
                }
            } else {
                break;
            }
        }
    }
}

// fetch the anagrams of a word -- useful for --strict and --prove
fn words_in_word(word: &str, noah: Arc<AnagramFun>, threads: usize) -> Vec<String> {
    let mut cc = noah
        .root
        .translator
        .count(&normalize(""))
        .expect("no luck with the char count");
    if let Some(usizes) = noah.root.translator.translate(word) {
        if !cc.add(usizes) {
            dictionary_error(word, &noah.root)
        }
    } else {
        dictionary_error(word, &noah.root)
    }
    cc.set_limits();
    let materials = vec![ToDo::seed(cc)];
    let (messages, _) = factory::manufacture(threads, 3, materials, noah.clone());
    let mut words = vec![];
    for m in messages {
        if let Some(todo) = m {
            for word in todo.words() {
                words.push(noah.root.translator.etalsnart(&word).unwrap());
            }
        } else {
            break;
        }
    }
    words
}

fn dictionary_error(word: &str, trie: &Trie) -> ! {
    let (good, bad) = trie.translator.unfamiliar_character(word);
    eprintln!(
        "character in {} not present in any word in dictionary:\n\n\t{}-->{}",
        word, good, bad
    );
    process::exit(1)
}

fn make_trie(opts: &ArgMatches, minimum_word_length: usize) -> Trie {
    let mut file = match File::open(opts.value_of("dictionary").unwrap()) {
        Err(_) => {
            eprintln!("could not read dictionary:\n\n{}", opts.usage());
            process::exit(1)
        }
        Ok(file) => file,
    };
    let mut strings = String::new();
    match file.read_to_string(&mut strings) {
        Err(why) => {
            eprintln!(
                "could not read words from dictionary: {}\n\n{}",
                why,
                opts.usage()
            );
            process::exit(1)
        }
        Ok(_) => (),
    }
    let words: Vec<&str> = strings
        .lines()
        .filter(|w| w.trim().len() >= minimum_word_length)
        .collect();
    let translator = Translator::new(normalize, words.iter().map(|s| *s));
    let any_excluded = opts.is_present("exclude");
    let mut excluded = HashSet::new();
    if any_excluded {
        for word in opts.values_of("exclude").unwrap() {
            excluded.insert(translator.translate(word).unwrap());
        }
    }
    let mut t = TrieNodeBuilder::new();
    for word in words {
        let translation = translator.translate(word).unwrap();
        if any_excluded && excluded.contains(&translation) {
            continue;
        }
        t.add(&translation);
    }
    let random = opts.is_present("random");
    Trie::new(
        t.build(),
        translator,
        !(opts.is_present("no_cache") || opts.is_present("set")),
        random,
    )
}

struct AnagramFun {
    root: Trie,
}

impl WorkerFun<ToDo> for AnagramFun {
    fn improve(&self, needs_work: ToDo) -> Vec<ToDo> {
        let mut done = vec![];
        let arc = Arc::new(needs_work);
        for (word, cc) in self.root.words_for(arc.undone.clone(), &arc.word, &false) {
            done.push(ToDo::new(arc.clone(), word, cc.clone()))
        }
        done
    }
    fn inspect(&self, thing: &ToDo) -> bool {
        thing.done()
    }
}


src/cli.rs
//! command line specification and usage information

extern crate clap;
use self::clap::{App, Arg};

pub fn parse<'a>(cpus: &'a str, dictionary: Option<&'a str>) -> App<'a, 'a> {
    let mut dictionary_argument = Arg::with_name("dictionary")
        .short("d")
        .long("dictionary")
        .value_name("file")
        .help("A line-delimited list of words usable in anagrams")
        .takes_value(true);
    if let Some(file) = dictionary {
        dictionary_argument = dictionary_argument.default_value(file);
    }
    App::new("rana")
        .version(crate_version!())
        .author(crate_authors!(", "))
        .about(crate_description!())
        .arg(dictionary_argument)
        .arg(
            Arg::with_name("set")
                .short("w")
                .long("words-in")
                .display_order(1)
                .help("Returns the set of words composable from the letters in the input phrase"),
        )
        .arg(
            Arg::with_name("strict")
                .long("strict")
                .requires("set")
                .display_order(2)
                .help("When finding --words-in, returns only words that occur in some anagram"),
        )
        .arg(
            Arg::with_name("prove")
                .long("prove")
                .requires("set")
                .display_order(3)
                .help("Like --strict, but emits a phrase proving this word occurs in an anagram."),
        )
        .arg(
            Arg::with_name("include")
                .short("i")
                .long("include")
                .value_name("word")
                .help("Include this word in the anagrams")
                .takes_value(true)
                .empty_values(false)
                .multiple(true)
                .number_of_values(1),
        )
        .arg(
            Arg::with_name("exclude")
                .short("x")
                .long("exclude")
                .value_name("word")
                .help("Exclude this word from anagrams")
                .takes_value(true)
                .empty_values(false)
                .multiple(true)
                .number_of_values(1),
        )
        .arg(
            Arg::with_name("threads")
                .short("t")
                .long("threads")
                .takes_value(true)
                .empty_values(false)
                .default_value(cpus)
                .value_name("n")
                .help("The number of threads to use during anagram collection"),
        )
        .arg(
            Arg::with_name("limit")
                .short("l")
                .long("limit")
                .takes_value(true)
                .empty_values(false)
                .value_name("n")
                .help("Only find this many anagrams")
                .conflicts_with("set"),
        )
        .arg(
            Arg::with_name("min")
                .short("m")
                .long("minimum-word-length")
                .takes_value(true)
                .empty_values(false)
                .value_name("n")
                .help("Words in anagrams must be at least this long"),
        )
        .arg(
            Arg::with_name("phrase")
                .value_name("word")
                .multiple(true)
                .required_unless_one(&["long-help", "ribbit"])
                .help("The words for which you want an anagram"),
        )
        .arg(
            Arg::with_name("no_cache")
                .short("C")
                .long("no-cache")
                .help("Do not cache partial results (this saves memory and costs speed)"),
        )
        .arg(
            Arg::with_name("random")
                .short("r")
                .long("random")
                .help("(Partially) shuffle order of discovery"),
        )
        .arg(
            Arg::with_name("long-help")
                .long("help-long")
                .help("Prints *detailed* help information"),
        )
        .arg(Arg::with_name("ribbit").long("ribbit").help("Ego sum"))
}

pub fn long_help() -> String {
    r#"Rana generates all the possible anagrams from a given phrase and
dictionary. Note "given some dictionary." Rana does not have a word list
built in. You must tell it what words it may use in an anagram. I have made
myself such a list out of a list of English words I found on the Internet from
which I delted all the words likely to offend people. By default rana will
look in your home directory for a file called .anagrams-dictionary.txt.

In many cases a simple phrase will have hundreds of thousands or millions of
anagrams, setting aside permutations. The phrase "rotten apple", for example,
with a fairly ordinary dictionary of of 109,217 English words, produces 2695
anagrams. Here are 10:

    pone prattle
    plea portent
    pole pattern
    portent pale
    platter pone
    planter poet
    potent paler
    porn palette
    pron palette
    poler patent

Because so many anagrams are available, you are likely to want to focus your
search. Rana provides several options to facilitate this.

--words-in

This will list all the words in your dictionary composable from some subset of
your phrase. You likely want to add --strict to this, so you only get words that
occur in *some* anagram. The --strict version is slower. If you want to verify
that each word listed has some anagram, you can add --prove, or replace --strict
with --prove. This will cause rana to emit the remaining words in an anagram
using the word in question.

--exclude

Discard from your word list particular words.

--include

Include only those phrases which include particular words.

--limit

Only provide a sample of this many phrases.

--random

Shuffle the search order over partial results while searching for anagrams. This
does not provide a fully random sample of the possible anagrams, since only the
results found at any point are shuffled, not all possible results, but this is
a decent way to look at a sample of anagrams when the phrase you've fed in has
many thousands of results. This is particularly useful when paired with --limit.

Caching and Threads

Rana by default uses as many processing threads as there are cores on your
machine. Generally this is what you want, but if you've got a lot of other
things going on, you can limit the number of available threads to reduce the
load your kernel has to deal with.

Rana also uses a dynamic programming algorithm to reduce the complexity of
finding algorithms for large phrases. This is probably unnecessary for short
phrases, though rana provides no lower limit. For larger phrases, like the
complete alphabet, the cache used by the dynamic programming algorithm may grow
so large that the process crashes. If you turn off the cache rana will use
a constant amount of memory, though it may take considerably longer to find all
anagrams.

Text Normalization

Rana attempts to strip away certain characters from your word list and all
other textual input, so it will treat "c-a-t" and " C A T " the same as "cat".
Here is the actual code that does this:

    pub fn normalize(word: &str) -> String {
        word.trim()
            .to_lowercase()
            .chars()
            .filter(|c| c.is_alphabetic())
            .collect::<String>()
    }

I have not tested what this will do for something like ß or Í. You may want to
normalize the text yourself before you give it to rana.

NOTE:

The caching algorithm treats character counts as long base-10 numbers. So, for
example, "cat" might be 111 if "c" is 100s, "a" is 10s, and "t" is 1s. Then
"cad" might be 1110 -- "d" is 1000s, and there is one of them, but there is no
"t", so the 1s place is 0. If you have a phrase, like

    "Dhrtaraashtra uvaaca, dharmakshetre kurukshetre samavetaa yuyutsavah"

which has 15 a's, the "a" count has to be represented as 5 -- 15 mod 10 is 5. In
other words, there isn't space in the a's column of the number for all the a's
in the phrase. It is unlikely that this will ever cause trouble, because for a
particular phrase you are unlikely to be encounter two sets of character counts
during processing that have different counts but the same code. Also, a phrase
this size will consume so much cache space that the process will probably crash
before you encounter this collision.

Another consideration with caching is that this scheme can only accommodate
alphabets up to 38 characters in size.
"#
    .to_string()
}
